translate/storage/markdown.py

```diff
@@ -0,0 +1,462 @@
+#
+# Copyright 2023 Anders Kaplan
+#
+# This file is part of the Translate Toolkit.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>.
+
+"""Module for parsing Markdown files for translation.
+
+The principles for extraction of translation units are as follows:
+
+1. Extract all content relevant for translation, at the cost of also
+   including some formatting.
+2. One translation unit per paragraph.
+3. Keep formatting out of the translation units as much as possible.
+   Exceptions include *phrase emphasis* and `inline code`.
+   Use placeholders {1}, {2}, ..., as needed.
+4. Avoid HTML entities in the translation units. Use Unicode
+   equivalents if possible.
+
+White space within translation units is normalized, because the PO format does
+not preserve white space, and the translated Markdown content may have
+to be reflowed anyway.
+"""
+import re
+from itertools import chain
+from typing import Iterable, Tuple
+
+import mistletoe.block_token as block_token
+import mistletoe.span_token as span_token
+import mistletoe.token
+from mistletoe.markdown_renderer import (
+    Fragment,
+    LinkReferenceDefinition,
+    LinkReferenceDefinitionBlock,
+    MarkdownRenderer,
+)
+
+from translate.storage import base
+
+
+class mdunit(base.TranslationUnit):
```

Comment:
Please consider using consistent naming with markdownfile. Using MarkdownUnit / MarkdownFile might be even better. Unfortunately, current storage classes are inconsistent, but I think it is better to stick with PEP-8 for new code.

```suggestion
class MarkdownUnit(base.TranslationUnit):
```